File: app/src/main/java/me/leoyuu/opengl/tmp/Render.kt
package me.leoyuu.opengl.tmp

import android.opengl.GLES30.*
import android.opengl.GLSurfaceView
import me.leoyuu.opengl.glsl.createProgram
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer

import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

/**
 * date 2018/12/8
 * email july.oloy@qq.com
 *
 * @author leoyuu
 */
class Render : GLSurfaceView.Renderer {
    private var vPosition = 0
    private var vColor = 0
    private var p = 0
    private var vao = IntArray(1)
    private var vbo = IntArray(1)

    private val vertices = floatArrayOf(
        0.5f,   0.5f,
        -0.5f, -0.5f,
        0.5f,  -0.5f
    )

    private val vertexBuf: FloatBuffer

    init {
        val vbb = ByteBuffer.allocateDirect(vertices.size * 4)
        vbb.order(ByteOrder.nativeOrder())
        vertexBuf = vbb.asFloatBuffer()
        vertexBuf.put(vertices)
        vertexBuf.position(0)
    }

    override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
        glClearColor(1.0f, 0.0f, 0.0f, 1.0f)
        p = createProgram("tmp/tmp.vert", "tmp/tmp.frag")
        vPosition = glGetAttribLocation(p, "vPosition")
        vColor = glGetUniformLocation(p, "vColor")

        glUniform4f(vColor, 0.0f, 1.0f, 0.0f, 1.0f)
        glGenVertexArrays(1, vao, 0)
        glBindVertexArray(vao[0])
        glGenBuffers(1, vbo, 0)
        glBindBuffer(GL_ARRAY_BUFFER, vbo[0])
        glBufferData(GL_ARRAY_BUFFER, vertices.size * 4, vertexBuf, GL_STATIC_DRAW)
        glEnableVertexAttribArray(vPosition)
        glVertexAttribPointer(vPosition, 2, GL_FLOAT, false, 0, 0)
    }

    override fun onSurfaceChanged(gl: GL10?, width: Int, height: Int) {
        glViewport(0, 0, width, height)
    }

    override fun onDrawFrame(gl: GL10) {
        glClear(GL_DEPTH_BUFFER_BIT or GL_COLOR_BUFFER_BIT)
        glUseProgram(p)
        glBindVertexArray(vao[0])
        glDrawArrays(GL_TRIANGLE_STRIP, 0, 3)
        glBindVertexArray(0)
    }

}

File: app/src/main/cpp/draw/renders/Render.cpp
//
// Created by 刘煜 on 2018/12/10.
//

#include <android/log.h>
#include <random>
#include "Render.h"
#include "../gl_helper.h"

#define LOG_TAG "render"

#define  LOG_I(...)  __android_log_print(ANDROID_LOG_INFO, LOG_TAG, __VA_ARGS__)
#define  LOG_E(...)  __android_log_print(ANDROID_LOG_ERROR,LOG_TAG, __VA_ARGS__)

#define G_VERTICES_SHADER  "tmp/tmp.vert"
#define G_FRAGMENT_SHADER  "tmp/tmp.frag"


const uint VERTICES_NUMBER = 5;
const GLfloat gTriangleVertices[] = { 0.0f,  0.5f,
                                      -0.5f, -0.5f,
                                      0.5f, -0.5f,
                                      -1.0f, -1.0f,
                                      1.0f, -1.0f};

bool setupGraphics(GLuint program) {
    printGLString("Version", GL_VERSION);
    printGLString("Vendor", GL_VENDOR);
    printGLString("Renderer", GL_RENDERER);
    printGLString("Extensions", GL_EXTENSIONS);

    glClearColor(0.0f, 1.0f, 1.0f, 1.0f);

    GLuint vPosition= (GLuint)glGetAttribLocation(program, "vPosition");
    checkGlError("glGetAttribLocation");

    LOG_I("glGetAttribLocation(\"vPosition\") = %d\n", vPosition);
    GLuint names[1];
    glGenBuffers(1, names);
    checkGlError("glGenBuffers");
    glBindBuffer(GL_ARRAY_BUFFER, names[0]);
    checkGlError("glBindBuffer");
    glBufferData(GL_ARRAY_BUFFER, 4 * 2 * VERTICES_NUMBER, gTriangleVertices, GL_STATIC_DRAW);
    checkGlError("glBufferData");
    glVertexAttribPointer(vPosition, 2, GL_FLOAT, GL_FALSE, 0, 0);
    checkGlError("glVertexAttribPointer");

    return true;
}


void drawFrame(GLuint program) {
    checkGlError("glClearColor");
    glClear( GL_DEPTH_BUFFER_BIT | GL_COLOR_BUFFER_BIT);
    checkGlError("glClear");

    glUseProgram(program);
    checkGlError("glUseProgram");
    GLuint vPosition= (GLuint)glGetAttribLocation(program, "vPosition");

    checkGlError("glVertexAttribPointer");
    glEnableVertexAttribArray(vPosition);
    checkGlError("glEnableVertexAttribArray");
    glDrawArrays(GL_TRIANGLE_STRIP, 0, VERTICES_NUMBER);
    checkGlError("glDrawArrays");
    glDisableVertexAttribArray(vPosition);
    checkGlError("glDisableVertexAttribArray");
}

Render::Render() {
    LOG_I("new render");
}

void Render::init() {
    this->program = createProgram(G_VERTICES_SHADER, G_FRAGMENT_SHADER);
    setupGraphics(this->program);
}

void Render::resize(uint width, uint height) {
    glViewport(0, 0, width, height);
}

void Render::renderFrame() {
    drawFrame(this->program);
}

void Render::click() {
    LOG_I("click");
}

Render::~Render() {
    LOG_I("delete render");
}